pageserver/src/storage_sync.rs

```diff
@@ -129,1330 +114,587 @@
+                "remote task {} completed successfully after {} retries",
+                task.op, retries
+            );
+        } else {
+            info!("remote task {} completed successfully", task.op);
+        }
 
-/// bool in return value stands for awaits_download
-fn compare_local_and_remote_timeline(
-    new_sync_tasks: &mut VecDeque<(TenantTimelineId, SyncTask)>,
-    sync_id: TenantTimelineId,
-    local_metadata: TimelineMetadata,
-    local_files: HashSet<PathBuf>,
-    remote_entry: &RemoteTimeline,
-) -> (LocalTimelineInitStatus, bool) {
-    let _entered = info_span!("compare_local_and_remote_timeline", sync_id = %sync_id).entered();
-
-    let remote_files = remote_entry.stored_files();
-
-    let number_of_layers_to_download = remote_files.difference(&local_files).count();
-    let (initial_timeline_status, awaits_download) = if number_of_layers_to_download > 0 {
-        new_sync_tasks.push_back((
-            sync_id,
-            SyncTask::download(LayersDownload {
-                layers_to_skip: local_files.clone(),
-            }),
-        ));
-        info!("NeedsSync");
-        (LocalTimelineInitStatus::NeedsSync, true)
-        // we do not need to manipulate with remote consistent lsn here
-        // because it will be updated when sync will be completed
-    } else {
-        info!("LocallyComplete");
-        (
-            LocalTimelineInitStatus::LocallyComplete(local_metadata.clone()),
-            false,
-        )
-    };
-
-    let layers_to_upload = local_files
-        .difference(remote_files)
-        .cloned()
-        .collect::<HashSet<_>>();
-    if !layers_to_upload.is_empty() {
-        new_sync_tasks.push_back((
-            sync_id,
-            SyncTask::upload(LayersUpload {
-                layers_to_upload,
-                uploaded_layers: HashSet::new(),
-                metadata: Some(local_metadata),
-            }),
-        ));
-        // Note that status here doesn't change.
-    }
+        // The task has completed succesfully. Remove it from the in-progress list.
+        {
+            let mut upload_queue = self.upload_queue.lock().unwrap();
+            upload_queue.inprogress_tasks.remove(&task.task_id);
 
-    (initial_timeline_status, awaits_download)
+            match task.op {
+                UploadOp::UploadLayer(_) => {
+                    upload_queue.num_inprogress_layer_uploads -= 1;
+                }
+                UploadOp::UploadMetadata(_, lsn) => {
+                    upload_queue.num_inprogress_metadata_uploads -= 1;
+                    upload_queue.last_uploaded_consistent_lsn = Some(lsn);
+                }
+                UploadOp::Delete(_) => {
+                    upload_queue.num_inprogress_deletions -= 1;
+                }
+                UploadOp::Barrier(_) => unreachable!(),
+            };
+
+            // Launch any queued tasks that were unblocked by this one.
+            self.launch_queued_tasks(upload_queue);
```

Comment:
To double check: there's no background job that retries concurrent tasks, they are run only when the corresponding function (`wait_completion`, `schedule_*`, recursively for dependent tasks).

In case any task fails, the whole dependent chain waits for another request? Not sure how optimal is that, but some comment is worth to be present for that.

Also potentially, that spawns a lot of `task_mgr::spawn` tasks without any global registry or even cancellations(?) now. Seems like another big refactoring is better to be done to make task management more observable, wdyt?

Reply:
>To double check: there's no background job that retries concurrent tasks, they are run only when the corresponding function (wait_completion, schedule_*, recursively for dependent tasks).

We launch a separate task for each upload operation, and that task is responsible for retrying it until it succeeds. `perform_upload_task' is infallible.

> In case any task fails, the whole dependent chain waits for another request? Not sure how optimal is that, but some comment is worth to be present for that.

If an upload fails, the task will keep retrying until it succeeds. Any other operations in the queue that are blocked by that operation will indeed wait until it succeeds. For example, if you schedule an upload of a layer file, and after that the upload of the index file, and the layer-file upload fails, the index file upload is blocked until it is retried and succeeds. If there are any other uploads after the index-file upload, they are blocked too, there's no "jumping the queue".

You're right that this is simpler than the logic we have in 'main' currently. On 'main', if the upload of one layer file fails, we would retry it later, but in the meanwhile we would try to upload later layer files from the queue. That's good in that you can make some progress. It's more complicated, however, and I don't think it's really needed. Until all the layer uploads have finished, you cannot upload the index file anyway, and without the index file, the files in S3 won't actually be visible or usable by anyone. And the uploading isn't latency-critical, it's OK to have some delay there.

> Also potentially, that spawns a lot of task_mgr::spawn tasks without any global registry or even cancellations(?) now. Seems like another big refactoring is better to be done to make task management more observable, wdyt?

`task_mgr.rs` *is* a global registry. Yeah, we might need some mechanism to cancel the tasks on timeline deletin or detach. Currently, we'll just have to wait for all the operations in the queue to finish.

@problame added a prometheus metric for the number of unfinished tasks in the queue, so that's something already.